lib/ObjWriter/typeBuilder.cpp

```diff
@@ -0,0 +1,124 @@
+#include "typeBuilder.h"
+#include "llvm/Support/COFF.h"
+
+UserDefinedTypesBuilder::UserDefinedTypesBuilder()
+    : Allocator(), TypeTable(Allocator) {}
+
+void UserDefinedTypesBuilder::SetStreamer(MCObjectStreamer *Streamer) {
+  assert(this->Streamer == 0);
+  assert(Streamer != 0);
+  this->Streamer = Streamer;
+}
+
+void UserDefinedTypesBuilder::EmitCodeViewMagicVersion() {
+  Streamer->EmitValueToAlignment(4);
+  Streamer->AddComment("Debug section magic");
+  Streamer->EmitIntValue(COFF::DEBUG_SECTION_MAGIC, 4);
+}
+
+ClassOptions UserDefinedTypesBuilder::GetCommonClassOptions() {
+  return ClassOptions();
+}
+
+void UserDefinedTypesBuilder::EmitTypeInformation(
+    MCSection *COFFDebugTypesSection) {
+
+  if (TypeTable.empty())
+    return;
+
+  Streamer->SwitchSection(COFFDebugTypesSection);
+  EmitCodeViewMagicVersion();
+
+  TypeTable.ForEachRecord([&](TypeIndex FieldTypeIndex,
+                              ArrayRef<uint8_t> Record) {
+    StringRef S(reinterpret_cast<const char *>(Record.data()), Record.size());
+    Streamer->EmitBinaryData(S);
+  });
+}
+
+unsigned UserDefinedTypesBuilder::GetEnumFieldListType(
+    uint64 Count, EnumRecordTypeDescriptor *TypeRecords) {
+  FieldListRecordBuilder FLRB(TypeTable);
+  FLRB.begin();
+  uint64 MaxInt = (unsigned int)-1;
+  if (Count > MaxInt) {
+    // Do not expect more than int fields.
+    return 0;
```

Comment:
This looks like it should be surfaced as an error instead of silent return of 0, at least in debug mode.